w3af/plugins/crawl/url_payment_finder.py

```diff
@@ -0,0 +1,219 @@
+the Free Software Foundation version 2 of the License.
+
+w3af is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with w3af; if not, write to the Free Software
+Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+
+"""
+from itertools import chain, repeat, izip
+
+import w3af.core.controllers.output_manager as om
+import w3af.core.data.kb.knowledge_base as kb
+from w3af.core.controllers.plugins.crawl_plugin import CrawlPlugin
+from w3af.core.controllers.core_helpers.fingerprint_404 import is_404
+from w3af.core.data.dc.headers import Headers
+from w3af.core.data.bloomfilter.scalable_bloom import ScalableBloomFilter
+from w3af.core.data.parsers.doc.url import URL
+from w3af.core.data.kb.info import Info
+from w3af.core.data.request.fuzzable_request import FuzzableRequest
+
+
+class url_payment_finder(CrawlPlugin):
+    """
+    Try to find hidden gateway payment confirmation page.
+    :author: Coiffey Pierre (pierre.coiffey@gmail.com)
+    """
+    _dirs = ('/', '/inc/', '/include/', '/include/pay/', '/includes/', '/includes/pay/', '/lib/',
+             '/libraries/', '/module/', '/module/pay/', '/modules/', '/modules/pay/', '/payment/',
+             '/shop/', '/store/', '/svc/', '/servlet/', '/cgi/', '/cgi-bin/', '/cgibin/',
+    )
+
+    _files = ('pay', 'payment', 'success', 'paymentsuccess', 'paymentcomplete', 'paymentsuccessful',
+              'successful', 'paid', 'return', 'valid', 'validpay', 'validate', 'validatepayment',
+              'validatepay', 'validation', 'complete', 'completepay', 'completepayment',
+              'trxcomplete', 'transactioncomplete', 'final', 'finished',
+    )
+    _exts = ('', '.php', '.asp', '.aspx', '.jsp', '.py', 
+             '.pl', '.rb', '.cgi', '.php3',  '.php4', '.php5',
+    )
+
+    def __init__(self):
+
+        CrawlPlugin.__init__(self)
+
+        self._headers = None
+        self._first_time = True
+        self._fuzz_images = False
+        self._seen = ScalableBloomFilter()
+
+    def crawl(self, fuzzable_request):
+        """
+        Searches for new Url's using fuzzing.
+
+        :param fuzzable_request: A fuzzable_request instance that contains
+                                    (among other things) the URL to test.
+        """
+        url = fuzzable_request.get_url()
+        self._headers = Headers([('Referer', url.url_string)])
+
+        if self._first_time:
+            self._verify_head_enabled(url)
+            self._first_time = False
+
+        # First we need to delete fragments and query strings from URL.
+        url = url.uri2url()
+
+        # And we mark this one as a "do not return" URL, because the
+        # core already found it using another technique.
+        self._seen.add(url)
```

Comment:
This doesn't look OK.

If the framework calls this plugin with `http://host/path?id=1` and `http://host/path?id=2` then the plugin will send duplicated HTTP requests to:

 * `http://host/pay`
 * `http://host/payment`
 * `http://host/success`
 * ...

Right?

What we want to do is:
 * Transform the input URI to a URL
 * Check if we already performed the analysis on that URL. If we did, then return
 * Perform the analysis on the URL, because it was never done before

Reply:
Still an issue

Reply:
Yes I agree, the modifications should be ok.